packages/flutter_web_plugins/lib/flutter_web_plugins.dart

```diff
@@ -15,6 +15,8 @@
 ///    describing how the `url_launcher` package was created using [flutter_web_plugins].
 library flutter_web_plugins;
 
+export 'src/navigation/js_url_strategy.dart';
 export 'src/navigation/url_strategy.dart';
+export 'src/navigation/utils.dart';
```

Comment:
Is it possible to keep `js_url_strategy.dart` and `utils.dart` unexported? `url_strategy.dart` exposes enough API for users.

Reply:
By policy we always export everything. Private APIs should be underscore-prefixed. Having a "private" API that is just not exported doesn't work, people can (and do) still import those library explicitly. If we think these APIs should be private I can submit a second PR that uses underscore-prefixing to limit their visibility, though.

Reply:
@Hixie That's good to know, thanks for clarifying. I prefer to keep them private and not expose unnecessary APIs to users. I'm not sure how to do that without placing everything in one file though (since `url_strategy.dart` calls into `js_url_strategy.dart` and `utils.dart`). Are you talking about underscore-prefixing file names? Or underscore-prefixing classes?

Reply:
Putting everything in one file is what I meant, yeah.

These files are pretty short, it would probably be fine to do that if you feel strongly about wanting these APIs private.

Reply:
> By policy we always export everything. [...] people can (and do) still import those library explicitly. [...] Putting everything in one file is what I meant

@Hixie may I add a +1 on Mouad's push-back on this?

I see a problem with exporting private APIs like this: they'll pollute the documentation of the library with information that the user doesn't need to know/care about.

Moving everything to a big file (be it by copy-pasting into a big file, or using a `part` directives) makes the library harder to reason about and maintain (at least, for me).

While migrating old plugins to the new Federated architecture, I found some maintainability issues derived from the `_` pattern (mostly in `part`ed libraries, and especially with _PrivateClasses and _PrivateEnums):

* Code is harder to read. Depending on how much 'private' API exists, a file might end up growing to a huge size. See any web plugin for examples where most (lots) of the code is private, and we need to tightly control what gets exported and what doesn't.
* Code is harder to test. IIRC `_PrivateThings` can't be seen/used from your unit tests, (can they???)
* Code is harder to extend/refactor/split after it's been merged. For example: how can a private `_Enum` be shared from several library entrypoints if it's not in a common file? What if `_Enum` then needs to become public because it needs to be shared across packages?

The `part` directive (and I guess, by extension, copy-pasting code into a single big file) is discouraged by Dart in: "[Creating packages: Organizing a library package](https://dart.dev/guides/libraries/create-library-packages#organizing-a-library-package)":

>> **Note:** You may have heard of the `part` directive, which allows you to split a library into multiple Dart files. We recommend that you avoid using `part` and create mini libraries instead.

WRT importing from somebody else's `src`; according to Dart's "[Package layout conventions: Implementation files](https://dart.dev/tools/pub/package-layout#implementation-files)" (emphasis mine): 

>> You are free to import libraries that live in `lib/src` from within other Dart code in the same package (like other libraries in `lib`, scripts in `bin`, and tests) but **you should never import from another package’s `lib/src` directory. Those files are not part of the package’s public API, and they might change in ways that could break your code.**

I understand that people may have imported from `lib/src` on occasion, but they really **shouldn't**. (Can we make [the `implementation_imports` lint](https://dart-lang.github.io/linter/lints/implementation_imports.html) mandatory at the dart-sdk level? :P)

Reply:
I agree with everything you say. The problem is that in practice people find "hidden" APIs and there's a psychological thing where using a hidden/undocumented API feels "cool", and so then they're using a bunch of undocumented APIs and the APIs not being documented becomes a hindrance, not a benefit... and we don't get any of the benefits of "internal" APIs being easier to break.

This is why our breaking change policy doesn't take public/private into account. It doesn't matter if an API we want to change is documented or not; what matters is whether or not it breaks a test in flutter/tests (the customer_testing shard).

That's why I just made it all public. The reality is that they are all public anyway.

Reply:
If I had to make the choice, I'd prefer "all public" than "all in a single file" too :P